File: cpp/tests/abstractions/crates.cpp
#include "types/string.hpp"

#include <cassert>

using namespace rust::serde_json;
using namespace rust::crate::abstractions;

int main()
{
	auto json_str_rust = R"({"name": "John", "age": 30})"_rs;
	auto json_rust_result = from_str(json_str_rust);
	assert(json_rust_result.is_ok());

	auto json_rust = json_rust_result.unwrap();
	remove_key_from_json(json_rust, "age"_rs);

	auto after_json_string_rust_result = to_string(json_rust);
	assert(after_json_string_rust_result.is_ok());
	auto after_json_string_rust = after_json_string_rust_result.unwrap();
	auto after_json_string_cpp = intoCpp<std::string>(std::move(after_json_string_rust));
	assert(after_json_string_cpp == R"({"name":"John"})");
}

File: rust/src/abstractions/crates.rs
use serde_json::Value;

pub fn remove_key_from_json(json_obj: &mut Value, key: &str) {
    if let Value::Object(map) = json_obj {
        map.remove(key);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::{from_str, to_string};

    #[test]
    fn test_string_to_json() {
        let json_str = r#"{"name": "John", "age": 30}"#;
        let json_obj: Result<Value, _> = from_str(json_str);
        assert!(json_obj.is_ok());
        assert_eq!(
            json_obj.unwrap(),
            serde_json::json!({"name": "John", "age": 30})
        );
    }

    #[test]
    fn test_json_to_string() {
        let json_obj = serde_json::json!({"name": "John", "age": 30});
        let json_str = to_string(&json_obj);
        assert!(json_str.is_ok());
        let result = json_str.unwrap();
        assert!(result == r#"{"name":"John","age":30}"# || result == r#"{"age":30,"name":"John"}"#);
    }

    #[test]
    fn test_remove_key_from_json() {
        let mut json_obj = serde_json::json!({"name": "John", "age": 30});
        remove_key_from_json(&mut json_obj, "age");
        assert_eq!(json_obj, serde_json::json!({"name": "John"}));
    }
}
